Add API V2 from DMPonline

gjacob24 opened the pull request:
PR for issue #3586 

This issue branch uses the doorkeeper gem to implement the Authorization Code Grant Flow (ACGF) part of the v2 API.
(The Client Credentials Flow (CCF) part of the v2 API (which is what the org-admins will use) will be implemented separately in a future piece of work.)

gjacob24 committed:
Add API V2 code from DMPonline service

martaribeiro committed:
add doorkeeper to routes

aaronskiba commented:
Thank you for this PR.

During today's DMPRoadmap Monthly all-team meeting I mentioned how there are also related DMPTool v2 API test files. Would you like me to add them to this PR?

momo3404 commented:
I'm seeing a lot of rubocop errors in [here](https://github.com/DMPRoadmap/roadmap/actions/runs/19576717664/job/56063943239?pr=3587). This might be a difference in the configuration of Rubocop between the Roadmap repo and DMP Online. Let me know how you'd want to approach this. If possible, I could start a PR attempting to fix these.

aaronskiba commented on app/views/api/v2/datasets/_show.json.jbuilder:
```diff
@@ -0,0 +1,84 @@
+
+  json.type output.output_type
+  json.title output.title
+  json.doi_url output.doi_url
```
It looks like `output.doi_url` will raise `*** NoMethodError Exception: undefined method 'doi_url' for #<ResearchOutput ...`.

Does DMPonline have a `doi_url` method defined? If not, I'm also seeing the following code comment in `app/services/external_apis/doi_service.rb`:
  ```ruby
  def mint(plan:)
    SecureRandom.uuid

    # Minted DOIs should be stored as an Identifier. For example:
    #    doi_url = "#{landing_page_url}#{doi}"
    #    Identifier.new(identifiable: plan, value: doi_url)

    # When this service is active and the above identifier is available,
    # the link to the DOI will appear on the Project Details page, in plan
    # exports and will become the `dmp_id` in this system's API responses
  end
  ```

gjacob24 replied:
Hi @aaronskiba, apologies I missed this. Not sure why it didn't throw an error when I tested it. Please remove line 10 ` json.doi_url output.doi_url` from `app/views/api/v2/datasets/_show.json.jbuilder` because it's from the DOI in research outputs feature that we added in DMPonline. The DOI feature was meant to be shared back to roadmap by now in a different PR, that's why it's missing.

Regarding the file `app/services/external_apis/doi_service.rb`, this is not part of the API V2 code. It's part of something Brian worked on for DOI minting.

martaribeiro commented:
Here is the document with the API calls. 
[API_v2.pdf](https://github.com/user-attachments/files/23826928/API_v2.pdf)

Please let us know if needs changing

gjacob24 commented:
> I'm seeing a lot of rubocop errors in [here](https://github.com/DMPRoadmap/roadmap/actions/runs/19576717664/job/56063943239?pr=3587). This might be a difference in the configuration of Rubocop between the Roadmap repo and DMP Online. Let me know how you'd want to approach this. If possible, I could start a PR attempting to fix these.

@momo3404 apologies we missed this. I can do this and push it back next week, if that's ok.

martaribeiro commented:
These two docs are more for developers support. It might be useful 
[Authorization_Code_Grant_Flow.pdf](https://github.com/user-attachments/files/23827
[doorkeeper_gem_for_OAuth_2.0_flows.pdf](https://github.com/user-attachments/files/23827180/doorkeeper_gem_for_OAuth_2.0_flows.pdf)
178/Authorization_Code_Grant_Flow.pdf)

aaronskiba commented:
> These two docs are more for developers support. It might be useful [Authorization_Code_Grant_Flow.pdf](https://github.com/user-attachments/files/23827 [doorkeeper_gem_for_OAuth_2.0_flows.pdf](https://github.com/user-attachments/files/23827180/doorkeeper_gem_for_OAuth_2.0_flows.pdf) 178/Authorization_Code_Grant_Flow.pdf)

Awesome, thank you. Sorry, the links got a little bungled up in the comment. :P
https://github.com/user-attachments/files/23827180/doorkeeper_gem_for_OAuth_2.0_flows.pdf works.
Could you just provide the Authorization_Code_Grant_Flow.pdf link again?

gjacob24 committed:
Add rubocop fixes

gjacob24 commented:
> > These two docs are more for developers support. It might be useful [Authorization_Code_Grant_Flow.pdf](https://github.com/user-attachments/files/23827 [doorkeeper_gem_for_OAuth_2.0_flows.pdf](https://github.com/user-attachments/files/23827180/doorkeeper_gem_for_OAuth_2.0_flows.pdf) 178/Authorization_Code_Grant_Flow.pdf)
> 
> Awesome, thank you. Sorry, the links got a little bungled up in the comment. :P https://github.com/user-attachments/files/23827180/doorkeeper_gem_for_OAuth_2.0_flows.pdf works. Could you just provide the Authorization_Code_Grant_Flow.pdf link again?

@aaronskiba Here you go:
[Authorization_Code_Grant_Flow.pdf](https://github.com/user-attachments/files/23853339/Authorization_Code_Grant_Flow.pdf)

Let us know if this doesn't work.

aaronskiba commented:
I'm encountering the following when I test the `GET /api/v2/heartbeat` endpoint (I can also see that the endpoint is returning a 500 for DMPOnline):

```
$ curl http://127.0.0.1:3000/api/v2/heartbeat
NoMethodError at /api/v2/heartbeat
==================================

undefined method `name' for nil:NilClass

> To access an interactive console with this error, point your browser to: /__better_errors


app/views/api/v2/_standard_response.json.jbuilder, line 18
----------------------------------------------------------

``` ruby
   13   json.ignore_nil!
   14   
   15   json.server @server
   16   json.source "#{request.method} #{request.path}"
   17   json.time Time.now.to_formatted_s(:iso8601)
>  18   json.client @client.name
   19   json.code response.status
   20   json.message Rack::Utils::HTTP_STATUS_CODES[response.status]
   21   
   22   if response.status == 200
   23   
```

App backtrace
-------------

 - app/views/api/v2/_standard_response.json.jbuilder:18:in `_app_views_api_v___standard_response_json_jbuilder__1455696549825964062_81440'
 - app/views/api/v2/error.json.jbuilder:3:in `_app_views_api_v__error_json_jbuilder__3591388735008983784_81420'
 - app/controllers/api/v2/base_api_controller.rb:69:in `handle_internal_server_error'
 - app/controllers/api/v2/base_api_controller.rb:58:in `handle_exception'

Full backtrace
...
```

aaronskiba commented on db/migrate/20251119131055_create_doorkeeper_tables.rb:
```diff
@@ -0,0 +1,98 @@
+      t.string  :uid,     null: false
+      t.string  :secret,  null: false
+
+      # Remove `null: false` if you are planning to use grant flows
+      # that doesn't require redirect URI to be used during authorization
+      # like Client Credentials flow or Resource Owner Password.
+      t.text    :redirect_uri, null: false
```
Since we will be implementing the v2 API Client Credentials Flow in a future piece of work, maybe we should remove `null: false` on redirect_uri,  as the code comment advises.

aaronskiba commented on app/controllers/api/v2/templates_controller.rb:
```diff
@@ -0,0 +1,19 @@
+      respond_to :json
+
+      # GET /api/v2/templates
+      def index
```
There is no route defined for this in `config/routes.rb`. I think we need something like `resources :templates, only: :index` inside of `namespace :v2 do`.

aaronskiba commented on app/views/api/v2/error.json.jbuilder:
```diff
@@ -0,0 +1,7 @@
+
+# json.items []
+json.message @payload[:message]
+json.details @payload[:details]
```
I can see `@payload[:message]` being assigned within `app/controllers/api/v2/base_api_controller.rb`. But is `@payload[:details]` being set anywhere? Maybe it can be removed.

aaronskiba commented on app/controllers/api/v2/base_api_controller.rb:
```diff
@@ -0,0 +1,94 @@
+        @scopes = doorkeeper_token.scopes.to_a if doorkeeper_token
+        return unless doorkeeper_token&.resource_owner_id
+
+        @resource_owner = User.find(doorkeeper_token.resource_owner_id)
```
I guess this could be refactored. An outer `if doorkeeper_token` could be used to replace the repeated `if doorkeeper_token` checks.

gjacob24 replied:
@johnpinto1 and I looked into this method today and have updated it. The endpoint can now be accessed by anyone as opposed to it's previous state. Let us know if this is ok.

aaronskiba commented on app/models/oauth_application.rb:
```diff
@@ -0,0 +1,18 @@
+# created_at:         :datetime
+# updated_at:         :datetime
+
+class OauthApplication < ApplicationRecord
```
I'm a little worried about directly inheriting from `ApplicationRecord` and simply mapping to the Doorkeeper tables in `app/models/oauth_access_grant.rb`, `app/models/oauth_access_token.rb`, and `app/models/oauth_application.rb`. These classes will not include Doorkeeper’s behaviour (validations, associations, etc.).

Currently, `@client = OauthApplication.find(doorkeeper_token.application_id) if doorkeeper_token` appears to be the only code that references these models (see `app/controllers/api/v2/base_api_controller.rb`), which should be safe. But I worry about issues arising in the future.

Unless this was implemented for a specific reason, I might suggest one of the following:
1. Remove these classes altogether
- These tables/models can instead be referenced via `Doorkeeper::Application`, `Doorkeeper::AccessToken;`, and `Doorkeeper::AccessGrant`. (e.g. `@client = Doorkeeper::Application.find(doorkeeper_token.application_id) if doorkeeper_token`)
2. Update each of the classes to inherit from the appropriate `Doorkeeper::Application`, `Doorkeeper::AccessToken;`, or `Doorkeeper::AccessGrant`.
  ```ruby
  class OauthApplication   < Doorkeeper::Application; end
  class OauthAccessToken   < Doorkeeper::AccessToken; end
  class OauthAccessGrant   < Doorkeeper::AccessGrant; end
  ```

aaronskiba commented on app/policies/api/v2/plans_policy.rb:
```diff
@@ -0,0 +1,31 @@
+        def resolve
+          Plan.joins(:roles)
+              .where(roles: { user_id: @resource_owner.id, active: true })
+              .distinct
```
It might be even better to reuse the `Plan.active(@resource_owner)` scope for this query. Note that a bit of refactoring could be performed in that scope to make this more efficient (i.e. don't use `includes(:template, :roles)`)

momo3404 commented:
Please let me know if I've done something wrong here, but I keep getting 302s when executing the GET command from the Authorization Code Grant Flow. 
`curl -i -X GET "http://127.0.0.1:3000/oauth/authorize?response_type=code&client_id=[my_client_id]&redirect_uri=[redirect_uri]"
HTTP/1.1 302 Found
x-frame-options: SAMEORIGIN
x-xss-protection: 0
x-content-type-options: nosniff
x-permitted-cross-domain-policies: none
referrer-policy: strict-origin-when-cross-origin
location: http://127.0.0.1:3000/users/sign_in
content-type: text/html; charset=utf-8
cache-control: no-cache`

But the command still worked, because after calling `Doorkeeper::AccessGrant.first` I found my authorization which I used for the POST command, and that worked perfectly. So is this a known error, or something I'm missing? I was also signed in when making the request.

aaronskiba commented on app/views/api/v2/templates/index.json.jbuilder:
```diff
@@ -0,0 +1,25 @@
+    end
+
+    json.template_id do
+      identifier = Api::V2::ConversionService.to_identifier(context: @application,
```
I think `@application` will evaluate to nil here.
I can see that app/controllers/api/v1/base_api_controller.rb sets `@application = ApplicationService.application_name`, whereas app/controllers/api/v2/base_api_controller.rb is setting `@server = ApplicationService.application_name`.

aaronskiba committed:
Add v2 API test coverage

This test suite was adapted from the v2 API tests in CDLUC3/dmptool.

The original request and view specs were copied and then refactored to align with our v2 API. Supporting factories and spec helper files were updated as needed to accommodate the new coverage.

aaronskiba committed:
Fix bugs uncovered by v2 API test coverage

These changes resolve the following issues uncovered via the API v2 tests:

- app/views/api/v2/datasets/_show.json.jbuilder
  - Failure rendering output.doi_url when output is present (undefined method `doi_url`)
- config/routes.rb
  - Spec failures due to undefined route helpers in test context (e.g., `api_v2_templates_path`)

gjacob24 committed:
Remove DMPonline specific code from doorkeeper initializer

Co-Authored-By: gjacob24 <97518971+gjacob24@users.noreply.github.com>

gjacob24 commented:
> I'm encountering the following when I test the `GET /api/v2/heartbeat` endpoint (I can also see that the endpoint is returning a 500 for DMPOnline):
> 
> ```
> $ curl http://127.0.0.1:3000/api/v2/heartbeat
> NoMethodError at /api/v2/heartbeat
> ==================================
> 
> undefined method `name' for nil:NilClass
> 
> > To access an interactive console with this error, point your browser to: /__better_errors
> 
> 
> app/views/api/v2/_standard_response.json.jbuilder, line 18
> ----------------------------------------------------------
> 
> ``` ruby
>    13   json.ignore_nil!
>    14   
>    15   json.server @server
>    16   json.source "#{request.method} #{request.path}"
>    17   json.time Time.now.to_formatted_s(:iso8601)
> >  18   json.client @client.name
>    19   json.code response.status
>    20   json.message Rack::Utils::HTTP_STATUS_CODES[response.status]
>    21   
>    22   if response.status == 200
>    23   
> ```
> 
> ## App backtrace
> 
>     * app/views/api/v2/_standard_response.json.jbuilder:18:in `_app_views_api_v___standard_response_json_jbuilder__1455696549825964062_81440'
> 
>     * app/views/api/v2/error.json.jbuilder:3:in `_app_views_api_v__error_json_jbuilder__3591388735008983784_81420'
> 
>     * app/controllers/api/v2/base_api_controller.rb:69:in `handle_internal_server_error'
> 
>     * app/controllers/api/v2/base_api_controller.rb:58:in `handle_exception'
> 
> 
> Full backtrace ...

@aaronskiba this is fixed and should be working now

aaronskiba commented on app/views/doorkeeper/applications/_form.html.erb:
```diff
@@ -0,0 +1,42 @@
+<%= form_for application, url: doorkeeper_submit_path(application), as: :doorkeeper_application, html: { role: 'form' } do |f| %>
```
We should update the admin panel for superusers so that they can navigate to this `oauth/applications` page without having to manually type the path in the browser.


<img width="1290" height="506" alt="Screenshot from 2025-12-22 11-44-12" src="https://github.com/user-attachments/assets/23e6acc8-0d43-42b2-9fd4-66148262f49e" />

<img width="231" height="453" alt="Screenshot from 2025-12-22 11-36-58" src="https://github.com/user-attachments/assets/2811ce7b-d030-408b-9790-fa5c9bead3ac" />

aaronskiba commented:
> > Full backtrace ...
> 
> @aaronskiba this is fixed and should be working now

Cool. I can get the 200 via the heartbeat whether I'm signed in or signed in now.

Unfortunately, I think the fix may have introduced other bugs. I am testing the following curl request:
```
$ curl -H "Authorization: Bearer ${BEARER_TOKEN}"      -H "Accept: application/json"      "http://127.0.0.1:3000/api/v2/plans"
```

This works and returns the expected JSON response on https://github.com/DMPRoadmap/roadmap/pull/3589, which is not up-to-date with these lastest heartbeat changes. However, when I test on this PR, I get the following response:

```
{
  "source": "GET /api/v2/plans",
  "time": "2025-12-22T12:03:00-07:00",
  "code": 500,
  "message": [
    "There was a problem in the server."
  ]
```
I also get the following in the rails console:
```
Processing by Api::V2::PlansController#index as JSON
  Doorkeeper::AccessToken Load (1.3ms)  SELECT "oauth_access_tokens".* FROM "oauth_access_tokens" WHERE "oauth_access_tokens"."token" = $1 LIMIT $2  [["token", "lp5znDvJAOZndbJRWAZ1eeZ_5NR8ZZs_2raswR-YsgQ"], ["LIMIT", 1]]
Exception message: undefined method `include?' for nil:NilClass

        raise Pundit::NotAuthorizedError unless @scopes.include?('read')
                                                       ^^^^^^^^^
  Rendering api/v2/error.json.jbuilder
  Rendered api/v2/_standard_response.json.jbuilder (Duration: 0.7ms | Allocations: 400)
  Rendered api/v2/error.json.jbuilder (Duration: 1.8ms | Allocations: 728)
Completed 500 Internal Server Error in 10ms (Views: 3.8ms | ActiveRecord: 0.9ms | Allocations: 7697)
```
This must be related to `@scopes = doorkeeper_token.scopes.to_a if doorkeeper_token` being removed in commit 3645cae2f7dcfa36437dc618f46d12a132c2c9d5

I am also looking at the `@application = ApplicationService.application_name` change in `app/controllers/api/v2/base_api_controller.rb`. I like this renaming because it is consistent with `app/controllers/api/v1/base_api_controller.rb`. However, `json.server @server` still exists in `app/views/api/v2/_standard_response.json.jbuilder`. Changing it to `json.application @application` should fix things and again align with the api/v1 approach.

The assignments to `@client` and `@resource_owner` were removed in commit 3645cae2f7dcfa36437dc618f46d12a132c2c9d5. This will need to be addressed as well.

gjacob24 commented:
> > > Full backtrace ...
> > 
> > 
> > @aaronskiba this is fixed and should be working now
> 
> Cool. I can get the 200 via the heartbeat whether I'm signed in or signed in now.
> 
> Unfortunately, I think the fix may have introduced other bugs. I am testing the following curl request:
> 
> ```
> $ curl -H "Authorization: Bearer ${BEARER_TOKEN}"      -H "Accept: application/json"      "http://127.0.0.1:3000/api/v2/plans"
> ```
> 
> This works and returns the expected JSON response on #3589, which is not up-to-date with these lastest heartbeat changes. However, when I test on this PR, I get the following response:
> 
> ```
> {
>   "source": "GET /api/v2/plans",
>   "time": "2025-12-22T12:03:00-07:00",
>   "code": 500,
>   "message": [
>     "There was a problem in the server."
>   ]
> ```
> 
> I also get the following in the rails console:
> 
> ```
> Processing by Api::V2::PlansController#index as JSON
>   Doorkeeper::AccessToken Load (1.3ms)  SELECT "oauth_access_tokens".* FROM "oauth_access_tokens" WHERE "oauth_access_tokens"."token" = $1 LIMIT $2  [["token", "lp5znDvJAOZndbJRWAZ1eeZ_5NR8ZZs_2raswR-YsgQ"], ["LIMIT", 1]]
> Exception message: undefined method `include?' for nil:NilClass
> 
>         raise Pundit::NotAuthorizedError unless @scopes.include?('read')
>                                                        ^^^^^^^^^
>   Rendering api/v2/error.json.jbuilder
>   Rendered api/v2/_standard_response.json.jbuilder (Duration: 0.7ms | Allocations: 400)
>   Rendered api/v2/error.json.jbuilder (Duration: 1.8ms | Allocations: 728)
> Completed 500 Internal Server Error in 10ms (Views: 3.8ms | ActiveRecord: 0.9ms | Allocations: 7697)
> ```
> 
> This must be related to `@scopes = doorkeeper_token.scopes.to_a if doorkeeper_token` being removed in commit [3645cae](https://github.com/DMPRoadmap/roadmap/commit/3645cae2f7dcfa36437dc618f46d12a132c2c9d5)
> 
> I am also looking at the `@application = ApplicationService.application_name` change in `app/controllers/api/v2/base_api_controller.rb`. I like this renaming because it is consistent with `app/controllers/api/v1/base_api_controller.rb`. However, `json.server @server` still exists in `app/views/api/v2/_standard_response.json.jbuilder`. Changing it to `json.application @application` should fix things and again align with the api/v1 approach.
> 
> The assignments to `@client` and `@resource_owner` were removed in commit [3645cae](https://github.com/DMPRoadmap/roadmap/commit/3645cae2f7dcfa36437dc618f46d12a132c2c9d5). This will need to be addressed as well.

Hey @aaronskiba, so sorry about this. I have changed `@server` to `@application`, to keep it consistent with V1. I've also put back `@scopes` and `@resource_owner`. That was removed unintentionally, so thanks for spotting that. You might still get other errors because we haven't fixed and tested them all, but I think this might be the last change I push before the long break to avoid causing more errors :D 

Also, please feel free to commit any changes you feel are necessary (some of which you have already added as comments in this PR).

aaronskiba committed:
Execute `translation:sync` with yaml enabled

Executed `bundle exec rake translation:sync` while `config.disable_yaml = true`, allowing us to sync the doorkeeper translations.

aaronskiba commented:
It looks like we have to sync the translations to get the Dookeeper views rendering properly (maybe this is only an issue in dev mode?). We usually have yaml translations disabled, but they'll have to be enabled in this instance.

https://github.com/DMPRoadmap/roadmap/commit/8befc57475126d2e2cc3f9fa11230ff8a3cd611b does exactly that and it looks like it fixes the rendering (see screenshots below). If you'd like, I could cherrypick the commit onto this PR?

BEFORE
<img width="980" height="404" alt="Screenshot from 2025-12-23 08-32-15" src="https://github.com/user-attachments/assets/0cc01d88-3f19-42bf-ac1a-80a85fd0dde7" />
AFTER
<img width="980" height="404" alt="Screenshot from 2025-12-23 08-32-30" src="https://github.com/user-attachments/assets/c23f8856-3654-4f18-aa73-62c4fdf0273c" />

Hiding this comment because [Execute translation:sync with yaml enabled](https://github.com/DMPRoadmap/roadmap/pull/3587/commits/6ed3d2d789407a3d00a25b2742e691f12c4f3231) has been applied to this PR.

aaronskiba committed:
Fix `@client` bug & rename server/client JSON keys

Fix crashes when `@client` is nil (e.g., on heartbeat endpoint where
doorkeeper_token is not present).
- `@client = doorkeeper_token&.application` simplifies `@client` assignment.
- Guard log_access and add `@caller` for safe JSON output

Align JSON response keys with V1
- server -> application
- client -> caller

Co-Authored-By: gjacob24 <97518971+gjacob24@users.noreply.github.com>
Co-Authored-By: John Pinto <8876215+johnpinto1@users.noreply.github.com>

aaronskiba committed:
Merge pull request #3589 from DMPRoadmap/aaron/add-api-v2-tests

Add v2 API Test Coverage and Address Uncovered Bugs

aaronskiba committed:
Add "OAuth Applications" to admin panel options

Add navigation link for Doorkeeper OAuth Applications management access in the admin dropdown menu.
- Link only renders for super admins.

aaronskiba committed:
Allow Doorkeeper HTTP redirects in test & dev

Allow HTTP OAuth redirects in test/dev for easier local development, while maintaining HTTPS enforcement in all other environments.

aaronskiba commented on app/policies/api/v2/plans_policy.rb:
```diff
@@ -0,0 +1,31 @@
+        end
+
+        def resolve
+          Plan.joins(:roles)
```
I'm getting a lot of N+1 query warnings from bullet. Maybe modifying the query here will help with that.

```
user: aaron
GET /api/v2/plans
USE eager loading detected
  Plan => [:roles]
  Add to your query: .includes([:roles])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:451:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Plan => [:contributors]
  Add to your query: .includes([:contributors])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:451:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Plan => [:identifiers]
  Add to your query: .includes([:identifiers])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:451:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Contributor => [:org]
  Add to your query: .includes([:org])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/contributors/_show.json.jbuilder:17:in `_app_views_api_v__contributors__show_json_jbuilder__813164984719583716_86260'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:43:in `block in _app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:42:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Contributor => [:identifiers]
  Add to your query: .includes([:identifiers])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/contributors/_show.json.jbuilder:17:in `_app_views_api_v__contributors__show_json_jbuilder__813164984719583716_86260'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:43:in `block in _app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:42:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Plan => [:funder]
  Add to your query: .includes([:funder])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:451:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Plan => [:research_outputs]
  Add to your query: .includes([:research_outputs])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:451:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Plan => [:template]
  Add to your query: .includes([:template])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:451:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Plan => [:org]
  Add to your query: .includes([:org])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:451:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'


user: aaron
GET /api/v2/plans
USE eager loading detected
  Role => [:user]
  Add to your query: .includes([:user])
Call stack
  /home/aaron/Documents/GitHub/roadmap_upstream/app/models/plan.rb:453:in `owner'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/presenters/api/v2/plan_presenter.rb:15:in `initialize'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `new'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/_show.json.jbuilder:7:in `_app_views_api_v__plans__show_json_jbuilder__4271210902697202284_86140'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:7:in `block (2 levels) in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:6:in `block in _app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/views/api/v2/plans/index.json.jbuilder:5:in `_app_views_api_v__plans_index_json_jbuilder___4045717872610860880_86120'
  /home/aaron/Documents/GitHub/roadmap_upstream/app/controllers/api/v2/plans_controller.rb:29:in `index'
```

aaronskiba commented on app/controllers/api/v2/base_api_controller.rb:
```diff
@@ -0,0 +1,97 @@
+      # set up pages in response
+      before_action :pagination_params, except: %i[heartbeat]
+
+      rescue_from Exception, with: :handle_exception
```
I'm guessing `rescue_from Exception, with: :handle_exception` is here to ensure API clients always get proper JSON error responses. However, `rescue_from StandardError, with: :handle_exception` might be better.

Rescuing from `Exception` will also catch system-level exceptions and that might be too far here. `StandardError` covers all application-level errors while still ensuring clients get consistent JSON responses.

aaronskiba committed:
Remove unneeded OAuth model wrappers

Remove OauthApplication, OauthAccessToken, and OauthAccessGrant model files.

These classes were empty wrappers around Doorkeeper models that inherited directly from ApplicationRecord, bypassing Doorkeeper's validations, associations, and behavior.

These wrapper classes were never referenced and posed a maintenance risk if used inadvertently in the future. Reference Doorkeeper's models directly instead (Doorkeeper::Application, Doorkeeper::AccessToken, and Doorkeeper::AccessGrant).

aaronskiba committed:
Fix assignment bug in fetch_q_and_a

aaronskiba committed:
Refactor `ResearchOutputPresenter#fetch_q_and_a`

`a.present? && !a.blank?` was redundant.
- Changed to simply `a.present?`

Replaces map + post-filter with filter_map
- Eliminates the need for `ret.select { |item| item[:description].present? }`

Extracts question-theme lookup and q & a formatting into private helper methods.
- Improves readability and helps address rubocop offences

aaronskiba committed:
Rescue `StandardError` in v2 BaseApiController

Rescuing from Exception would also catch system-level exceptions, which would likely be taking things too far (e.g. system-level errors like `Interrupt` and
`SystemExit`). StandardError covers all application-level errors while still ensuring clients get consistent JSON responses.

aaronskiba committed:
Update fields returned by GET /api/v2/me

Prior to this change, the `GET /me` endpoint exposed potentially sensitive user attributes (e.g. `api_token`) as well as  internal-only fields (e.g. `other_organisation`, `department_id`, `dmponline3`, etc.).

Now the endpoint only returns useful information (note the language.name rather than language_id and org.name rather than org_id improvements as well).

aaronskiba committed:
Refactor: Replace raw SQL in TemplatesPolicy

Replaced raw SQL with ActiveRecord query methods + enum scopes
- Improves readability and eliminates SQL injection risk

aaronskiba committed:
Filter PlanPresenter.plan_costs answers at db-level

Move theme filtering from Ruby to SQL using joins for better performance and scalability.

aaronskiba committed:
Refactor fetch_q_and_a: Filter answers at db level

Previously, `fetch_q_and_a` iterated over all plan questions and searched for matching answers in Ruby, performing theme filtering and Q&A formatting in memory. This caused unnecessary loops, potential N+1 lookups, and inefficient handling of multi-theme questions. Accessing questions directly via `Plan.questions` also triggered joins through sections, phases, and  templates, along with ordering (due to default scope), which added query overhead.

The refactored version:

- Queries answers directly via a join to question themes, filtering at the DB level.
- Accesses questions through the answers (`answer.question`) instead of `plan.questions`, avoiding heavy joins and unnecessary default-scoped sorting.

aaronskiba committed:
Optimise `GET api/v2/plans` / Fix N+1 queries

Previously, the /api/v2/plans endpoint triggered many N+1 queries when rendering plan associations, causing slow response times. This change adds .includes() for all associations that were previously triggering Bullet N+1 warnings.

Benchmark comparison (local dev, 10 sequential requests):

| Metric            | Before  | After  |
| ----------------- | ------- | ------ |
| Mean request time | 1909 ms | 492 ms |
| Requests/sec      | 0.52    | 2.03   |
| Max request time  | 2628 ms | 641 ms |

- This simple comparison shows that these changes make the  `GET api/v2/plans` ~4x faster.

aaronskiba committed:
Improve OAuth flow detection and redirect handling

- Use start_with?(oauth_authorization_path) instead of include?('/oauth/authorize') for stricter path matching
- Use stored_location_for(:user) instead of session[:user_return_to] for automatic cleanup and explicit Devise scope usage
  - OAuth flow detection is tied to session[:user_return_to], so the :user scope is explicit and correct
- Add ? suffix to predicate method name

Improves security (prevents substring matches), maintainability (uses framework APIs with explicit scope), and follows Ruby conventions.

aaronskiba committed:
Refactor & simply Doorkeeper `admin_authenticator`

Amongst its changes, this commit removes `warden.authenticate!(scope: :user)`.

When warden.authenticate! is invoked, the OAuth flow is preserved while the user signs in. However, our current custom implementation of ApplicationController#after_sign_in_path_for only returns `stored_location_for(resource)` for the `oauth/authorize` path.
- We could further modify after_sign_in_path_for to also return `stored_location_for(resource)` for the `oauth/applications` path. However, it is simple enough for a super admin to navigate to `oauth/applications` via the admin panel.

aaronskiba committed:
Simplify Doorkeeper authentication via Warden

- Use warden.authenticate! (instead of manually storing session[:user_return_to] + explicitly redirecting to login page)
- Simplify conditional logic by checking "native" path first
- Replace hardcoded "/oauth/authorize/native path" string with native_oauth_authorization_path

When warden.authenticate! is invoked, Warden both automatically stores the request URL in the session and reroutes the unauthenticated user to the login page.

The existing session[:user_return_to] checks in ApplicationController continue to work. This reduces custom code and leverages Devise's built-in authentication handling for better maintainability.

NOTE: For better alignment with Devise conventions and automatic session cleanup, we could refactor ApplicationController to use stored_location_for(resource) instead of session[:user_return_to].

aaronskiba committed:
Refactor: DRY up after_sign_in/after_sign_up helpers

With the exception of specific 'disallowed paths' (i.e. `new_user_session_path` for `after_sign_in_path_for` vs `new_user_session_path` and `new_user_registration_path` for `after_sign_up_path_for`, these helpers are otherwise identical. This change refactors the code so that these helpers better adhere to DRY principles.

We introduce `def after_auth_path(disallowed_paths:)`. `after_sign_in_path_for` and `after_sign_up_path_for` each call `after_auth_path` with an array containing their aforementioned disallowed paths. Some small additional refactoring is also performed. `return root_path if request.referer.nil? || from_external_domain?` allows for possible early return before assigning `referer_path`. Also, repeated `referer_path.eql?(path)` statements have been replaced with a single `disallowed_paths.include?(referer_path)` check.

momo3404 committed:
Add complete plan flag to `plans_controller`

- Edit plan query to check for user role to prevent eager loading
- Add set_complete_param function that sets complete flag according to its value if its set to true in the API call

momo3404 committed:
Add complete plan flag to `plan_presenter`

- Edit initialize to include complete flag, which is set to false as default
- If flag is true in call, call fetch_all_q_and_a
- Add fetch_all_q_and_a function that fetches questions and answers

Update plan_presenter

momo3404 committed:
Add complete plan flag to `plans/_show.json.jbuilder`

- Add complete flag data to extension in json

momo3404 committed:
Add question id to returned json

momo3404 committed:
Optimize complete plan query

- Pre-fetch the answers, questions, and sections as part of the initial plans query (but only when we are in fact fetching all of the questions and answers).
- This speeds up the mean request time by 25%

momo3404 committed:
Update CHANGELOG.md

momo3404 committed:
Merge pull request #3595 from DMPRoadmap/momo/add-complete-flag

Add complete plan flag to V2 API

aaronskiba committed:
Add Doorkeeper rake tasks to Rakefile

More information regarding these Doorkeeper rake tasks can be found at https://doorkeeper.gitbook.io/guides/internals/rake

aaronskiba committed:
Allow NULL `oauth_applications.redirect_uri` in db

Generate and apply migration to remove NOT NULL constraint from `oauth_applications.redirect_uri`.

This works in conjunction with `allow_blank_redirect_uri true` in `config/initializers/doorkeeper.rb`, allowing us to save `OAuthApplications` with blank `redirect_uri` values for URI-less flows like client_credentials or password grant flows.

NOTE: Applications using `authorization_code` (or other flows that require a redirect URI) must still have a non-blank `redirect_uri`. Otherwise, `/oauth/authorize` requests with `response_type=code` will fail.

More info is available here: https://github.com/doorkeeper-gem/doorkeeper/wiki/Allow-blank-redirect-URI-for-Applications

aaronskiba committed:
Enable Doorkeeper token/application secret hashing

- Enable `hash_application_secrets` and `hash_token_secrets` in Doorkeeper initializer for improved security.
- This stores application secrets and access tokens as hashes in the database, reducing risk if the database is compromised.
- Note: `hash_token_secrets` is incompatible with `reuse_access_token`, so token reuse is now disabled/removed (see warning in Doorkeeper docs).
- For more details, see: https://doorkeeper.gitbook.io/guides/security/token-and-application-secrets

aaronskiba committed:
Adapt v2 spec requests to now-hashed tokens

aaronskiba committed:
DRY up 'read' scope auth in v2 API controllers

Move the "read" scope authorization check to a shared `before_action` in `BaseApiController`
- Removes redundant checks from individual actions in `PlansController` and `TemplatesController`
- Allows for the safe removal of `@scopes` from `base_response_content`

aaronskiba committed:
Use `plan.owner.language` for `plan.language`

Prior to this change, `LocaleService.default_locale` was always assigned to `json.language`.
- Now we use `plan.owner&.language&.abbreviation` (or `LocaleService.default_locale` as a fallback)

aaronskiba committed:
Update plan.language test and V2 references

- Updated `it 'includes the :language'` test to reflect changes made in b238c8128e0fefb9ea25cc7528e5be2c499d9318.
- Updated `Api::V1` references within spec files with `Api::V2`

aaronskiba committed:
refactor: simplify `@data_contact` assignment

NOTE: The prior comment stated the following:
# Attach the first data_curation role as the data_contact, otherwise
# add the contributor to the contributors array

However, the contributor was ALWAYS added to the contributors array

aaronskiba committed:
refactor: use for_api_v2 scope for Template & Plan

Moved complex query logic from policies to `for_api_v2` scopes in Template and Plan models.

aaronskiba committed:
Eager load identifier_scheme in Template.for_api_v2

Addresses the following Bullet warning:

user: aaron
GET /api/v2/templates
USE eager loading detected
  Identifier => [:identifier_scheme]
  Add to your query: .includes([:identifier_scheme])
Call stack
  /home/aaron/Documents/GitHub/roadmap/app/presenters/api/v2/org_presenter.rb:9:in `block in affiliation_id'
  /home/aaron/Documents/GitHub/roadmap/app/presenters/api/v2/org_presenter.rb:9:in `affiliation_id'
  /home/aaron/Documents/GitHub/roadmap/app/views/api/v2/orgs/_show.json.jbuilder:11:in `block in _app_views_api_v__orgs__show_json_jbuilder__1237609272914133563_48080'
  /home/aaron/Documents/GitHub/roadmap/app/views/api/v2/orgs/_show.json.jbuilder:10:in `_app_views_api_v__orgs__show_json_jbuilder__1237609272914133563_48080'
  /home/aaron/Documents/GitHub/roadmap/app/views/api/v2/templates/index.json.jbuilder:16:in `block (3 levels) in _app_views_api_v__templates_index_json_jbuilder___4313931085157922975_47640'
  /home/aaron/Documents/GitHub/roadmap/app/views/api/v2/templates/index.json.jbuilder:15:in `block (2 levels) in _app_views_api_v__templates_index_json_jbuilder___4313931085157922975_47640'
  /home/aaron/Documents/GitHub/roadmap/app/views/api/v2/templates/index.json.jbuilder:8:in `block in _app_views_api_v__templates_index_json_jbuilder___4313931085157922975_47640'
  /home/aaron/Documents/GitHub/roadmap/app/views/api/v2/templates/index.json.jbuilder:5:in `_app_views_api_v__templates_index_json_jbuilder___4313931085157922975_47640'
  /home/aaron/Documents/GitHub/roadmap/app/controllers/api/v2/templates_controller.rb:13:in `index'

aaronskiba committed:
Optimise `GET api/v2/plans` / Fix N+1 queries

This change further addresses Bullet warnings that were earlier addressed in commit 6ed969beb62c6621271166a27a2854cc91b7b024

aaronskiba committed:
Use `.for_api_v2` scope for GET /api/v2/plans/:id

This change updates `Api::V2::PlansController#show` to use the `.for_api_v2` scope, and applies eager loading of answers (`.includes(answers: { question: :section })`) based on the `complete` param--making the `show` action consistent with `index`.

A new `plans_scope` helper DRYs up the controller by centralizing the shared scope and eager loading logic. Since `.for_api_v2` already filters by `.where(roles: { user_id: user_id, active: true })`, only a presence check is now required in the policy.
